File: loyautils/src/main/java/com/chadbingham/loyautils/rx/ScheduleFirebase.java
package com.chadbingham.loyautils.rx;

import java.util.concurrent.Executors;

import io.reactivex.CompletableTransformer;
import io.reactivex.FlowableTransformer;
import io.reactivex.MaybeTransformer;
import io.reactivex.ObservableTransformer;
import io.reactivex.Scheduler;
import io.reactivex.SingleTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class ScheduleFirebase {

	public static final Scheduler scheduler = Schedulers.from(Executors.newSingleThreadExecutor());

	public static <T> FlowableTransformer<T, T> flowable() {
		return o -> o.subscribeOn(scheduler).observeOn(AndroidSchedulers.mainThread());
	}

	public static <T> ObservableTransformer<T, T> observable() {
		return o -> o.subscribeOn(scheduler).observeOn(AndroidSchedulers.mainThread());
	}

	public static <T> SingleTransformer<T, T> single() {
		return o -> o.subscribeOn(scheduler).observeOn(AndroidSchedulers.mainThread());
	}

	public static <T> MaybeTransformer<T, T> maybe() {
		return o -> o.subscribeOn(scheduler).observeOn(AndroidSchedulers.mainThread());
	}

	public static CompletableTransformer completable() {
		return o -> o.subscribeOn(scheduler).observeOn(AndroidSchedulers.mainThread());
	}
}

File: loyautils/src/main/java/com/chadbingham/loyautils/fire/ScheduleFirebase.kt
package com.chadbingham.loyautils.fire

import io.reactivex.*
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.Executors

object ScheduleFirebase {

    private val scheduler = Schedulers.from(Executors.newSingleThreadExecutor())

    fun <T> flowable(): FlowableTransformer<T, T> {
        return FlowableTransformer { o -> o.subscribeOn(scheduler).observeOn(AndroidSchedulers.mainThread()) }
    }

    fun <T> observable(): ObservableTransformer<T, T> {
        return ObservableTransformer { o -> o.subscribeOn(scheduler).observeOn(AndroidSchedulers.mainThread()) }
    }

    fun <T> single(): SingleTransformer<T, T> {
        return SingleTransformer { o -> o.subscribeOn(scheduler).observeOn(AndroidSchedulers.mainThread()) }
    }

    fun <T> maybe(): MaybeTransformer<T, T> {
        return MaybeTransformer { o -> o.subscribeOn(scheduler).observeOn(AndroidSchedulers.mainThread()) }
    }

    fun completable(): CompletableTransformer {
        return CompletableTransformer { o -> o.subscribeOn(scheduler).observeOn(AndroidSchedulers.mainThread()) }
    }
}
